doc/mgmt/gnmi/gNMI_Subscription_for_YangData.md

```diff
@@ -0,0 +1,1183 @@
+  Following new translib APIs will be implemented:
+  - `IsSubscribeSupported()` to return subscription preferences for given paths.
+  - `Stream()` to return current data for given paths through a queue
+  - `Subscribe()` to handles ON_CHANGE subscriptions.
+    It starts monitoring the mapped DB key patterns using redis PSUBSCRIBE operations.
+    Subsequent changes to the DB will be stream asynchronously using a queue.
+
+- App modules are only required to provide YANG path to DB entry mappings and vice-versa.
+
+- App modules should implement app interface function `translateSubscribe()` to provide
+  YANG path to DB/non-DB mappings.
+  - Should include mappings for given YANG node and all its child nodes
+  - Should handle wildcard keys in the yang paths.
+  - Mapped DB key can have a redis wildcard pattern.
+  - One YANG path can have multiple DB/non-DB mappings.
+
+- App modules can also return subscription preferences in `translateSubscribe()` response.
+  Translib will assume default values if app did not specify the preferences.
+  Preferences includes:
+  - Whether ON_CHANGE enabled
+  - Minimum SAMPLE interval.
+  - Preferred subscription mode during TARGET_DEFINED
+
+- Transformer common_app will discover the mappings and preferences from the annotations.
+  Subtree transformers should implement `Subscribe_xxx()` callback function to specify
+  the mappings and preferences.
+  Table and key transformers should handle wildcard paths.
+
+- App modules should implement `processSubscribe()` app interface function to resolve YANG keys for a DB entry.
+  This is required to get a specific YANG path from a wildcard path template.
+
+- Transformer common_app will automatically support this for simple cases.
+  Apps should implement a *path transformer* to handle special cases --
+  subtree transformer or when there is no one-to-one mapping of YANG key to DB key.
+
+### 2.2 Identifying YANG Based Subscriptions
+
+Today the gNMI server supports subscription based on SONiC specific "DB paths" and "virtual paths"
+as described in the [GRPC Telemetry HLD](../../system-telemetry/grpc_telemetry.md).
+It uses the request `target` value to identify type of the subscribe paths.
+A subscribe request will be rejected if `target` was not specified.
+
+This is not in sync with the [gNMI specification](https://github.com/openconfig/reference/blob/master/rpc/gnmi/gnmi-specification.md#2221-path-target).
+`target` is optional and is supposed to be an opaque data for the server.
+However, the current behavior will not be removed completely for the sake of backward compatibility.
+The request `target` will be made optional.
+Request will be processed as described in [GRPC Telemetry HLD](../../system-telemetry/grpc_telemetry.md)
+if the `target` is specified and matches one of the reserved keywords listed in that HLD.
+If `target` is not specified or not one of the reserved ones, the subscribe path will be treated as tarnslib YANG path.
```

Comment:
@qiluo-msft , @zbud-msft , do we plan to use any other target, if yes, then this would break our future use cases.

Reply:
@lguohan & team -- did not hear any responses for this. Anyway, I now have proposed a `origin` based identification of OpenConfig YANG paths. Existing `target` based identification of DB/event/virtual paths will be retained when client did not specify an `origin`. Please check.
However this `target` based identification is a violation of the gNMI specification. Please consider using `origin` instead of `target`.